### PR TITLE
Add extreme-scale parallel solver

### PR DESCRIPTION
**Pull request by chengmiao0808**
Extreme-scale parallel linear programming (LP) solver can solve a large number of separate small LP problems in parallel based on the existing solver (which can only solve one LP at a time).

### CODE REVIEW

File: dualip/src/main/scala/com/linkedin/dualip/solver/ParallelLPSolverDriver.scala [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        )
+        val results = solve(objective, solver, initialLambda, driverParams, parallelMode = true)
+
+        (problemId, results.logList, results.dualList, results.violationList)
+      }.toDF("problemId", "logList", "dualList", "violationList")
+
+      resultsDS
+        .repartition(10)
+        .write
+        .format("avro")
```
Comment (panyao) [inline dualip/src/main/scala/com/linkedin/dualip/solver/ParallelLPSolverDriver.scala:91]: driverParam.outputFormat?

File: dualip/src/main/scala/com/linkedin/dualip/solver/MooDistributedRegularizedObjective.scala [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    val primalObjective = cx +  xx * gamma / 2.0
+    val slackMetadata: SlackMetadata = SolverUtility.getSlack(lambda.toArray, axMinusB.toArray, b.toArray)
+
+    // The sum of positive slacks, one of measures of constraints violation useful for logging
+    val absoluteConstraintsViolation =  axMinusB.toArray.filter(_ > 0.0).sum
+    log += ("dual_obj" -> f"$dualObjective%.8e")
+    log += ("cx" -> f"$cx%.8e")
+    log += ("feasibility" -> f"${slackMetadata.feasibility}%.6e")
+    log += ("λ(Ax-b)" -> f"${lambda dot gradient}%.6e")
+    log += ("γ||x||/2" -> f"${xx * gamma / 2.0}%.6e")
```
Comment (panyao) [inline dualip/src/main/scala/com/linkedin/dualip/solver/MooDistributedRegularizedObjective.scala:90]: γ||x||^2/2

File: dualip/src/main/scala/com/linkedin/dualip/problem/MooSolver.scala [hunk 3]
```diff
@@ -25,55 +25,67 @@
...
- *
- * Column (variable indices in "a" and "c" are relative, that is, variable is uniquely identified by
- * a combination of block id and internal id.
- *
- * Internal representation is optimized for the operations that algorithm implements and data characteristics:
- * in particular, dense constraints matrix with few rows.
+  * A MOO block of data. A vertical slice of design matrix, specifically the variables in the same simplex constraint sum x <= 1
+  * We need to keep this data together because we need to do a simplex projection on it.
+  *
+  * Column (variable0 indices in "a" and "c" are relative, that is, variable is uniquely identified by
```
Comment (panyao) [inline dualip/src/main/scala/com/linkedin/dualip/problem/MooSolver.scala:45]: variable0 -> variable

File: dualip/src/main/scala/com/linkedin/dualip/problem/MooSolver.scala [hunk 4]
```diff
@@ -137,27 +150,38 @@ class MooSolverDualObjectiveFunction(
...
+    * @param inputPaths input path for vectorB and ACblock
+    * @param spark spark session
+    * @return
+    */
+  def loadData(inputPaths: InputPaths)
+    (implicit spark: SparkSession): (MapReduceDataset[MooDataBlock], BSV[Double]) = {
     import spark.implicits._
 
     val budget = IOUtility.readDataFrame(inputPaths.vectorBPath, inputPaths.format)
-      .map{case Row(_c0: Number, _c1: Number) => (_c0.intValue(), _c1.doubleValue()) }
```
Comment (panyao) [inline dualip/src/main/scala/com/linkedin/dualip/problem/MooSolver.scala:160]: Can we keep this line? So the flow will not fail even user provide float input?

File: dualip/src/main/scala/com/linkedin/dualip/solver/LPSolverDriver.scala [hunk 5]
```diff
@@ -241,6 +242,41 @@ object LPSolverDriver {
...
+    implicit val spark = SparkSession
+      .builder()
+      .appName(getClass.getSimpleName)
+      .getOrCreate()
+
+    try {
+      val driverParams = LPSolverDriverParamsParser.parseArgs(args)
+
+      println("-----------------------------------------------------------------------")
+      println("                      Dualip v1.0     2021")
```
Comment (panyao) [inline dualip/src/main/scala/com/linkedin/dualip/solver/LPSolverDriver.scala:260]: DuaLip

File: dualip/src/main/scala/com/linkedin/dualip/problem/MooSolver.scala [hunk 6]
```diff
@@ -170,43 +194,85 @@ object MooSolverDualObjectiveFunction extends DualPrimalObjectiveLoader {
...
+}
+
+/**
+  * Parameters parser
+  */
+object MooParamsParser {
+  def parseArgs(args: Array[String]): MooObjectiveParams = {
+    val parser = new scopt.OptionParser[MooObjectiveParams]("Moo params parser") {
+      override def errorOnUnknownArgument = false
+      opt[Boolean]("moo.parallelMode") required() action { (x, c) => c.copy(parallelMode = x) }
```
[the remark this answers is not in the document]
Comment (chengmiao0808) [reply dualip/src/main/scala/com/linkedin/dualip/problem/MooSolver.scala:272]: sure, I'll make this parameter optional.

File: build.gradle [hunk 7]
```diff
@@ -26,7 +26,7 @@ allprojects {
 
     group = "com.linkedin.dualip"
     
-    project.version = "1.0.3"
+    project.version = "1.0.4"
```
[the remark this answers is not in the document]
Comment (chengmiao0808) [reply build.gradle:29]: Makes sense! As adding the parallel solver is a major change, will use the version number as 1.1.0.

File: dualip/src/main/scala/com/linkedin/dualip/problem/MooSolver.scala [hunk 8]
```diff
@@ -25,55 +25,67 @@
...
- * Internal representation is optimized for the operations that algorithm implements and data characteristics:
- * in particular, dense constraints matrix with few rows.
+  * A MOO block of data. A vertical slice of design matrix, specifically the variables in the same simplex constraint sum x <= 1
+  * We need to keep this data together because we need to do a simplex projection on it.
+  *
+  * Column (variable0 indices in "a" and "c" are relative, that is, variable is uniquely identified by
+  * a combination of block id and internal id.
+  *
+  * internal representation is optimized for the operations that algorithm implements and data characteristics:
+  * in particular, dense constraints matrix with few rows.
```
[the remark this answers is not in the document]
Comment (chengmiao0808) [reply dualip/src/main/scala/com/linkedin/dualip/problem/MooSolver.scala:49]: Sure

File: dualip/src/main/scala/com/linkedin/dualip/problem/MooSolver.scala (hunk 3 above)
Comment (chengmiao0808) [reply dualip/src/main/scala/com/linkedin/dualip/problem/MooSolver.scala:45]: thanks for the catch

File: dualip/src/main/scala/com/linkedin/dualip/solver/MooDistributedRegularizedObjective.scala (hunk 2 above)
Comment (chengmiao0808) [reply dualip/src/main/scala/com/linkedin/dualip/solver/MooDistributedRegularizedObjective.scala:90]: nice catch, will correct

**Pull request merged**